### ISSUE TITLE
Large amounts of repeated DB queries when using multiple federation sender workers

### ISSUE THREAD

**Issue by emelie-qis** [matrix-org/synapse#9113]
### Description

`wake_destinations_needing_catchup` performs massive amounts of repeated SELECT queries when using multiple federation_sender workers. Using a single federation_sender works well.

[Here](https://gist.github.com/emelie-qis/135264903b3778d117926873c2a9e449) are the debug logs from the federation senders, from startup up until the issue starts occuring. I don't think there's any value to providing more of the log, as the output just repeats itself after this point.

The actual destination in the query seems to vary though, I've seen three destinations so far: `toofat.ru`, `wc20.tencapsule.com` and `fpoe.info`. The output of a sample query with all three destinations follows below:

```
synapse_prod=# SELECT destination FROM destinations WHERE destination IN ( SELECT destination FROM destination_rooms WHERE destination_rooms.stream_ordering > destinations.last_successful_stream_ordering ) AND destination > 'toofat.ru' AND ( retry_last_ts IS NULL OR retry_last_ts + retry_interval < 1610625475512 ) ORDER BY destination LIMIT 25;
 destination
-------------
 wolffy.eu
(1 row)

synapse_prod=# SELECT destination FROM destinations WHERE destination IN ( SELECT destination FROM destination_rooms WHERE destination_rooms.stream_ordering > destinations.last_successful_stream_ordering ) AND destination > 'fpoe.info' AND ( retry_last_ts IS NULL OR retry_last_ts + retry_interval < 1610625676424 ) ORDER BY destination LIMIT 25;
     destination
---------------------
 komputernerds.com
 matrix.insw.cz
 matrix.noppesict.nl
 supersandro.de
 thedreamer.nl
 toofat.ru
 wolffy.eu
(7 rows)

synapse_prod=# SELECT destination FROM destinations WHERE destination IN ( SELECT destination FROM destination_rooms WHERE destination_rooms.stream_ordering > destinations.last_successful_stream_ordering ) AND destination > 'wc20.tencapsule.com' AND ( retry_last_ts IS NULL OR retry_last_ts + retry_interval < 1610631393252 ) ORDER BY destination LIMIT 25;
 destination
-------------
 wolffy.eu
(1 row)
```

### Steps to reproduce

- Enable two or more federation_sender workers
- Start the main process and all workers
- Wait ~30 seconds

Of course, this is probably specific to me and the contents of my db, so naturally I'll provide any further information as needed. Just let me know! :)

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

### Version information

<!-- IMPORTANT: please answer the following questions, to help us narrow down the problem -->

<!-- Was this issue identified on matrix.org or another homeserver? -->
- **Homeserver**: queersin.space

If not matrix.org:

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**: 1.25 (note: problem occurred on 1.24 as well)

- **Install method**: Via apt from the matrix.org repo
<!-- examples: package manager/git clone/pip  -->

- **Platform**: An lxc container running Ubuntu 20.04.1
<!--
Tell us about the environment in which your homeserver is operating
distro, hardware, if it's running in a vm/container, etc.
-->

**Comment by clokep** [comment]
Looks like we just noticed this as well -- it should be fixed by #9114.

**Comment by richvdh** [comment]
> Here are the debug logs 

the link seems to be broken?

>  performs massive amounts of repeated SELECT queries

why would the problem in #9114 cause lots of repeated queries (as opposed to a small number of very slow ones)?

**Comment by naosia** [comment]
Sorry, I deleted the logs as the issue was closed, in case there were any sensitive info in there I missed. It's back up now.

**Comment by richvdh** [comment]
Thanks.

TL;DR, there is a lot of this stuff going on:

```
021-01-14 14:51:48,760 - synapse.storage.SQL - 303 - DEBUG - None - [SQL time] {_find_stream_orderings_for_times_txn} 0.000124 sec
2021-01-14 14:51:48,761 - synapse.storage.SQL - 284 - DEBUG - None - [SQL] {_find_stream_orderings_for_times_txn} SELECT received_ts FROM events WHERE stream_ordering <= ? ORDER BY stream_ordering DESC LIMIT 1
2021-01-14 14:51:48,761 - synapse.storage.SQL - 289 - DEBUG - None - [SQL values] {_find_stream_orderings_for_times_txn} (2136523,)
2021-01-14 14:51:48,761 - synapse.storage.SQL - 303 - DEBUG - None - [SQL time] {_find_stream_orderings_for_times_txn} 0.000125 sec
2021-01-14 14:51:48,761 - synapse.storage.SQL - 284 - DEBUG - None - [SQL] {_find_stream_orderings_for_times_txn} SELECT received_ts FROM events WHERE stream_ordering <= ? ORDER BY stream_ordering DESC LIMIT 1
2021-01-14 14:51:48,761 - synapse.storage.SQL - 289 - DEBUG - None - [SQL values] {_find_stream_orderings_for_times_txn} (2065306,)
2021-01-14 14:51:48,761 - synapse.storage.SQL - 303 - DEBUG - None - [SQL time] {_find_stream_orderings_for_times_txn} 0.000119 sec
2021-01-14 14:51:48,761 - synapse.storage.SQL - 284 - DEBUG - None - [SQL] {_find_stream_orderings_for_times_txn} SELECT received_ts FROM events WHERE stream_ordering <= ? ORDER BY stream_ordering DESC LIMIT 1
2021-01-14 14:51:48,761 - synapse.storage.SQL - 289 - DEBUG - None - [SQL values] {_find_stream_orderings_for_times_txn} (2100915,)
2021-01-14 14:51:48,762 - synapse.storage.SQL - 303 - DEBUG - None - [SQL time] {_find_stream_orderings_for_times_txn} 0.000118 sec
2021-01-14 14:51:48,762 - synapse.storage.SQL - 284 - DEBUG - None - [SQL] {_find_stream_orderings_for_times_txn} SELECT received_ts FROM events WHERE stream_ordering <= ? ORDER BY stream_ordering DESC LIMIT 1
2021-01-14 14:51:48,762 - synapse.storage.SQL - 289 - DEBUG - None - [SQL values] {_find_stream_orderings_for_times_txn} (2083111,)
2021-01-14 14:51:48,762 - synapse.storage.SQL - 303 - DEBUG - None - [SQL time] {_find_stream_orderings_for_times_txn} 0.000116 sec
2021-01-14 14:51:48,762 - synapse.storage.SQL - 284 - DEBUG - None - [SQL] {_find_stream_orderings_for_times_txn} SELECT received_ts FROM events WHERE stream_ordering <= ? ORDER BY stream_ordering DESC LIMIT 1
2021-01-14 14:51:48,762 - synapse.storage.SQL - 289 - DEBUG - None - [SQL values] {_find_stream_orderings_for_times_txn} (2092013,)
2021-01-14 14:51:48,762 - synapse.storage.SQL - 303 - DEBUG - None - [SQL time] {_find_stream_orderings_for_times_txn} 0.000140 sec
2021-01-14 14:51:48,762 - synapse.storage.SQL - 284 - DEBUG - None - [SQL] {_find_stream_orderings_for_times_txn} SELECT received_ts FROM events WHERE stream_ordering <= ? ORDER BY stream_ordering DESC LIMIT 1
2021-01-14 14:51:48,762 - synapse.storage.SQL - 289 - DEBUG - None - [SQL values] {_find_stream_orderings_for_times_txn} (2087562,)
2021-01-14 14:51:48,763 - synapse.storage.SQL - 303 - DEBUG - None - [SQL time] {_find_stream_orderings_for_times_txn} 0.000122 sec
```

Again I'm not sure that's consistent with #9114 so I'm going to reopen this.

**Comment by erikjohnston** [comment]
Oh, right. This is it just iterating through all the destinations that need to be retried, but since there are multiple federation senders we rapidly chew through the list until we hit destinations we need to retry (I believe).

We may want to at some point clear out entries from the list that are e.g. more than a month old, to stop the list just ever increasing.

**Comment by emelie-qis** [comment]
Just to clarify *how much* of this stuff is going, I should add that the `wake_destinations_needing_catchup` job on each federation sender jumps to 100% DB use in the "DB usage by background jobs (including scheduling time)" metric.

**Comment by erikjohnston** [comment]
Yeah, we're seeing this on matrix.org as well. I'm hoping that changing the query means that it goes fast enough that we don't care too much that there are so many of them, but we'll see.

**Comment by emelie-qis** [comment]
Oh, I should've made that clear at the start - it keeps going indefinitely, I only noticed it yesterday morning after it had been going throughout the night. Sorry, it's my first time reporting an issue.

![2021-01-15-114733_1234x284_scrot](https://user-images.githubusercontent.com/614975/104715974-831d2b00-5727-11eb-93e5-c2d73fd7ed6f.png)

**Comment by emelie-qis** [comment]
Hi, just wanted to confirm that this was not related to #9114, and persists after upgrading to 1.26.0.

**Comment by clokep** [comment]
#9198 might help this a bit too, although the cause seems a bit different? (Note that that will be in 1.27.0 and requires using workers via Redis).

**Comment by emelie-qis** [comment]
This appears to be fixed in 1.27. Thanks a bunch for all your hard work! :smile:
